service/topology_coordinator.cc

```diff
@@ -66,6 +67,757 @@ future<inet_address> wait_for_ip(raft::server_id id, const raft_address_map& am,
+                if (r.host == target) {
+                    has_replica_on_target = true;
+                    break;
+                }
+                if (check_rack_load) {
+                    const locator::node& node = topo.get_node(r.host);
+                    if (node.dc_rack().dc == dc) {
+                        rack_load[node.dc_rack().rack] += 1;
+                    }
+                }
+            }
+        }
+
+        if (has_replica_on_target) {
+            stats.for_dc(dc).tablets_skipped_node++;
+            continue;
+        }
+
+        // Make sure we don't increase level of duplication of racks in the replica list.
+        if (check_rack_load) {
+            auto max_rack_load = std::max_element(rack_load.begin(), rack_load.end(),
+                                                [] (auto& a, auto& b) { return a.second < b.second; })->second;
+            auto new_rack_load = rack_load[target_node.dc_rack().rack] + 1;
+            if (new_rack_load > max_rack_load) {
+                stats.for_dc(dc).tablets_skipped_rack++;
+                continue;
+            }
+        }
+
+        auto& target_load_sketch = co_await target_info.get_load_sketch(tm);
+        auto dst = global_shard_id {target, target_load_sketch.next_shard(target)};
+
+        const locator::node& src_node = topo.get_node(src.host);
+        locator::tablet_transition_kind kind = (src_node.get_state() == locator::node::state::being_removed
+                                        || src_node.get_state() == locator::node::state::being_replaced)
+                    ? locator::tablet_transition_kind::rebuild : locator::tablet_transition_kind::migration;
+        auto mig = tablet_migration_info {kind, source_tablet, src, dst};
+        auto mig_streaming_info = get_migration_streaming_info(topo, tmap.get_tablet_info(source_tablet.tablet), mig);
+
+        if (can_accept_load(mig_streaming_info)) {
+            apply_load(mig_streaming_info);
+            stats.for_dc(dc).migrations_produced++;
+            new_replicas.emplace_back(dst);
+        } else {
+            // Shards are overloaded with streaming. Do not include the migration in the plan, but
+            // continue as if it was in the hope that we will find a migration which can be executed without
+            // violating the load. Next make_plan() invocation will notice that the migration was not executed.
+            // We should not just stop here because that can lead to underutilization of the cluster.
+            // Just because the next migration is blocked doesn't mean we could not proceed with migrations
+            // for other shards which are produced by the planner subsequently.
+            skipped_migrations++;
+            stats.for_dc(dc).migrations_skipped++;
+            if (skipped_migrations >= max_skipped_migrations) {
+                stats.for_dc(dc).stop_skip_limit++;
+                break;
+            }
+        }
```

Comment:
This is a copy of load balancer logic, most of which is not needed for replica allocation. We don't need load-limiting which throttles migrations, stats update, handling of node draining.